numpy/core/src/multiarray/descriptor.c

```diff
@@ -238,6 +238,22 @@ is_datetime_typestr(char const *type, Py_ssize_t len)
     return 0;
 }
 
+// This function checks if val is mistyped dict of inherited dtype,
+// e.g val = {key: tuple of size >= 2}; see gh-2865, issue #8235 for details
```

Comment:
I am a bit confused about what actually happens here. Could you give an example for when this function does *not* return `True` for `is_mistyped`? And how does such a descriptor differ from a mistyped one? What about the possibility that a metadata actually includes a correctly typed dtype?
So is it possible to create dtypes, e.g. with odd metadata, where this logic is broken?

Reply:
There two **contradicting** to each other usages of dictionary in a tuple that defines dtype.   Consider the following expression `np.dtype(('i4', adict))` (`'i4'` can be replaced by any other valid type):
 1. The dictionary `adict` here should be interpreted as `metadata`
     ```python
     dt =  np.dtype(('i4', {'comment': 'qty'}))
     dt.metadata
     # mappingproxy({'comment': 'qty'})

     ```
 1. But if we do this we will break historically working code `convert_from_iherit_tuple` (see #8235) that allows one to split a given type into parts and use up-to 2 names to access them. For example
     ```python
     dt = np.dtype(('i4', {'part_1': ('i2', 0, 'Part 1'), 'part_2': ('i2', 2, 'Part 2')}))
     dt.descr
     # [(('Part 1', 'part_1'), '<i2'), (('Part 2', 'part_2'), '<i2')]
     ```
     In this format tuple inside dictionary must be of size 2 or 3.

Thus, given a dictionary `adict` one has to decide if it is of type 1 or 2.  I've just submitted a change that makes this choice very clear. Specifically, `adict` is mistyped dict of type 2 if all of the following is true:
   1. all adict values are tuples of size >= 2;
   2. either of the following is true for at least one tuple
       a. val[1] is not an integer (this is required to pass existing regression tests);
       b. val[1] is an integer but val[1] < 0;
       c. len(val) >= 4.

I added description of this algo to docstring of the function `numpy.core._internal._is_mistyped_inherited_type_dict`.

Does this resolve your confusion?

Reply:
Yes, so what I need to figure out is whether we can deprecate that whole other meaning?

Because otherwise we are guessing what the user wants and guessing is bad. E.g.:
```
In [13]: dt = np.dtype(np.float64, metadata={"original_type": (np.int8, 8)})                                            
In [14]: np.dtype([("f1", dt)]).descr                                                                                   
Out[14]: [('f1', ('<f8', {'original_type': (numpy.int8, 8)}))]
```
In which case your code will misinterpret the metadata for a dtype when converting the `.descr`.
Sure, its a corner case, but if you want to build reliable software based on this guess, I feel the inconsistency will bite you at some point.

Reply:
I agree with you, reliability and consistency of software (especially widely used one) is very important .  It is hard decision and only one of you can make it.

Note that we are already in that inconsistent territory.  Recall my original concern: descr of dtype cannot be evaluated back into dtype when metadata is provided.

Reply:
So my feeling is that we should only do this, if it is reasonable to put a FutureWarning on the other branch (i.e. remove it). Even in that case, it would be nicer if we can put in the FutureWarning at least one version before going here.
But, I am not sure how valid the existing use-case is. It only works with `offsets` which I assume are not super-common, but if you use offsets, then it actually seems like a fairly inconsistent and nice way to spell this out.

I still would be a bit more curious about either dropping metadata entirely for `.descr` or just creating a new way to do this... maybe expose the buffer-protocol representation (which has its problems as well though).

Reply:
Another idea is to extend format of the tuple that  defines dtype: the tuple with metadata must have size 3.
1. `np.dtype((type, dict))` will be parsed as inherited dtype.
1. `np.dtype((type, None or dict_1, metadata)) will by parsed with metadata.

Reply:
I have a better idea that will keep the old code (and users of inherited dtype): require metadata to be stored in the dictionary under name metadata.  For example,
```python
np.dtype(('i4', {'part_1': ('i2', 0), 'part_2': ('i2', 2), 'metadata': {"x": 1}}))
np.dtype(('i8', {'metadata': {'x': 1}}))
np.dtype({'names': ['a', 'b'], 'formats': ['i2', 'i2'], 'metatdata': {'x': 1}})
```

In this way we can unambiguously infer dtype from dict: dict is either inherited dtype or dict with keys 'names', 'formats', 'offsets', ...

What do you think?

Reply:
Why do we need that when we have `np.dtype(type, metadata=...)` though?

Reply:
As far as I understand,
we want to be able to create dtype from a list and be able to specify metadata for individual fields, e.g.
   ```python
   np.dtype([('a', ('i8', {'x': 1})), ('b', ('f8', {'y': 1}))], metadata={'z': 1})
   ```
Keyword metadata only allows one to specify metadata for the final dtype.

The whole discussion appeared because adding metadata to dtype breaks construction of dtype from descr.